File: src/lib/asistencias-sync.ts
import { prisma } from "@/lib/prisma";
import { EstadoAsistencia } from "@prisma/client";

function eachDay(start: Date, end: Date): Date[] {
  const days: Date[] = [];
  const cur = new Date(start);
  cur.setHours(0, 0, 0, 0);
  const fin = new Date(end);
  fin.setHours(0, 0, 0, 0);
  while (cur <= fin) {
    days.push(new Date(cur));
    cur.setDate(cur.getDate() + 1);
  }
  return days;
}

export async function sincronizarAsistenciasSolicitud(
  empleadoId: number,
  fechaInicio: Date,
  fechaFin: Date,
  tipo: "VACACION" | "PERMISO"
) {
  const estado: EstadoAsistencia = tipo === "VACACION" ? "VACACION" : "PERMISO";
  const dias = eachDay(fechaInicio, fechaFin);

  for (const fecha of dias) {
    const dow = fecha.getDay();
    if (dow === 0 || dow === 6) continue;

    await prisma.asistencia.upsert({
      where: {
        empleadoId_fecha: {
          empleadoId,
          fecha,
        },
      },
      create: {
        empleadoId,
        fecha,
        estado,
        notas: `Generado por solicitud ${tipo}`,
      },
      update: {
        estado,
        notas: `Actualizado por solicitud ${tipo}`,
      },
    });
  }
}

File: backend/core/services/asistencias_sync.py
from datetime import timedelta

from django.utils import timezone

from core.models import Asistencia


def _each_day(start, end):
    cur = timezone.localtime(start).replace(hour=0, minute=0, second=0, microsecond=0)
    fin = timezone.localtime(end).replace(hour=0, minute=0, second=0, microsecond=0)
    days = []
    while cur <= fin:
        days.append(cur)
        cur = cur + timedelta(days=1)
    return days


def sincronizar_asistencias_solicitud(empleado_id: int, fecha_inicio, fecha_fin, tipo: str):
    estado = "VACACION" if tipo == "VACACION" else "PERMISO"
    now = timezone.now()
    for fecha in _each_day(fecha_inicio, fecha_fin):
        if fecha.weekday() >= 5:
            continue
        existing = Asistencia.objects.filter(empleado_id=empleado_id, fecha=fecha).first()
        if existing:
            Asistencia.objects.filter(id=existing.id).update(
                estado=estado,
                notas=f"Actualizado por solicitud {tipo}",
                updatedAt=now,
            )
        else:
            Asistencia.objects.create(
                empleado_id=empleado_id,
                fecha=fecha,
                estado=estado,
                notas=f"Generado por solicitud {tipo}",
                createdAt=now,
                updatedAt=now,
            )
